File: java/src/localsearchoptimization/components/Neighborhood.java
package localsearchoptimization.components;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class Neighborhood {

    protected Random random;

    protected Operator[] operators;

    protected List<Configuration> configurations = new ArrayList<Configuration>();

    protected Solution currentSolution;

    protected int currentConfiguration;

    public Neighborhood(Solution solution, Operator[] operators, int seed) {
        random = new Random(seed);
        currentSolution = solution;
        this.operators = operators;
        for (Operator operation : this.operators)
            configurations.addAll(operation.configurations);
        randomize();
    }

    public Solution getCurrentSolution() {
        return currentSolution;
    }

    public int getPower() {
        return configurations.size();
    }

    public Solution getNext() {
        return configurations.get(currentConfiguration++).apply(currentSolution);
    }

    public boolean hasNext() {
        return currentConfiguration < configurations.size();
    }

    public Solution getRandom() {
        return configurations.get(random.nextInt(getPower())).apply(currentSolution);
    }

    public void moveToSolution(Solution solution) {
        currentConfiguration = 0;
        currentSolution = solution;
    }

    public void randomize() {
        currentConfiguration = 0;
        Collections.shuffle(configurations, random);
    }

    public void reset() {
        currentConfiguration = 0;
    }
}

File: C#/LocalSearchOptimization/Components/Neighborhood.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace LocalSearchOptimization.Components
{
    public class Neighborhood
    {
        protected Random random;

        protected List<Operator> operators = new List<Operator>();

        protected List<Configuration> configurations = new List<Configuration>();

        public ISolution CurrentSolution { get; private set; }

        public int Power
        {
            get
            {
                return configurations.Count;
            }
        }

        public IEnumerable<ISolution> Neighbors
        {
            get
            {
                foreach (Configuration configuration in configurations)
                {
                    yield return configuration.Apply(CurrentSolution);
                }
            }
        }

        public Neighborhood(ISolution solution, List<Operator> operators, int seed)
        {
            random = new Random(seed);
            CurrentSolution = solution;
            this.operators.AddRange(operators);
            foreach (Operator operation in operators)
                configurations.AddRange(operation.Configurations);
            Randomize();
        }

        public virtual ISolution GetRandom()
        {
            return configurations[random.Next(Power)].Apply(CurrentSolution);
        }

        public void MoveToSolution(ISolution solution)
        {
            CurrentSolution = solution;
        }

        public void Randomize()
        {
            configurations = configurations.OrderBy(x => random.Next()).ToList();
        }
    }
}